Refactor Cookie, add CookieCollection

armpogart opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

Some preliminary refactoring and additional class (`CookieCoolection`) needed for #86 implementation.
Also I think changes are useful even without #86.

The `Cookie` class property defaults were changed to standard cookie defaults so that minimal valid cookie can be created (containing only name and empty value) and parsed by this class and recommended/useful defaults are now set in constructor with flag switch (enabled by default).

armpogart committed:
Refactor Cookie

rustamwin commented on src/Cookie.php:
```diff
@@ -1,14 +1,35 @@
  */
 final class Cookie
 {
+    // @see https://tools.ietf.org/html/rfc6265#section-4
```
Should be phpdoc

roxblnfk replied:
And it should be `@link` annotation instead of `@see`

xepozz commented on src/Cookie.php:
```diff
@@ -37,91 +58,103 @@ final class Cookie
      * @var string value of the cookie
      */
-    private string $value;
+    private string $value = '';
```
It shouldn't have default value

armpogart replied:
Yep, default value here can be omitted as constuctor sets it to default empty string anyway. Empty string is used instead of previous required value, as cookie can have empty value according to RFC.

xepozz commented on src/Cookie.php:
```diff
@@ -37,91 +58,103 @@ final class Cookie
-        $expireDateTime = (new \DateTimeImmutable())->add($dateInterval);
         $new = clone $this;
-        $new->expire = (int)$expireDateTime->format('U');
+        $new->expire = $dateTime->format('D, d M Y H:i:s T');
```
Would be nice to extract format template as a private constant

roxblnfk replied:
`\DateTimeInterface::RFC7231`

armpogart replied:
@roxblnfk I first used that constant, but it produces different result than the one defined in RFC, the result of using that constant is e.g. `Sun, 06 Nov 1994 08:49:37 +0000` but according to [RFC](https://tools.ietf.org/html/rfc2616#section-3.3.1) we need `Sun, 06 Nov 1994 08:49:37 GMT`. I couldn't find any information in RFC whether the first format will be ok, so I've used the custom format.

@xepozz Will extract it to constant with link to rfc specification.

armpogart replied:
@roxblnfk It occurs that `\DateTimeInterface::RFC7231` works correctly and tests just failed when I used it.

roxblnfk replied:
RFC7231 is the preferred date and time format in modern HTTP headers

xepozz commented on src/Cookie.php:
```diff
@@ -37,91 +58,103 @@ final class Cookie
     }
 
-    public function expireAt(\DateTimeInterface $dateTime): self
+    public function validFor(DateInterval $dateInterval): self
```
Is it really needed?

armpogart replied:
It is just a helper method. Not sure we need it or not as we can achieve the same with `expireAt`.

armpogart replied:
@xepozz I will rename the method to `maxAge` and it will set `Max-Age` header and also for compatibility reason and it will set `Expire` header to the same value as some agents/browsers don't support (read) `Max-Age` header, but it is present in specification. The behavior of the method will be the same.

xepozz commented on src/Cookie.php:
```diff
@@ -144,37 +177,131 @@ public function httpOnly(bool $httpOnly): self
+
+        if ($sameSite === self::SAME_SITE_NONE) {
+            // the secure flag is required for cookies that are marked as 'SameSite=None'
+            $this->secure = true;
```
Shouldn't you set `secure` to `$new` object?

armpogart replied:
Yep, my mistake will push a fix with further commit. The `Cookie` is immutable, so we don't change any property directly on it.

xepozz commented on src/Cookie.php:
```diff
@@ -37,91 +58,103 @@ final class Cookie
+        return $this->expireAt($expireDateTime);
+    }
+
+    public function expireWhenBrowserIsClosed(): self
```
Is it really needed? We have already `null` value when make new instance of Cookie

armpogart replied:
I see this one as a way to modify already existing cookie and removing expire header from it. This behavior can't be achieved by setting expire date to anything else. The only other way would be to create a new cookie, set all properties accordingly without setting expiry.

armpogart committed:
Add CookieCollection with tests, fixes to Cookie

armpogart committed:
Fix for styleci

armpogart committed:
Fix SessionMiddleware setting expire date

samdark commented on src/Cookie.php:
```diff
@@ -40,88 +71,120 @@ final class Cookie
+    private ?string $sameSite = null;
 
-    public function __construct(string $name, string $value)
+    public function __construct(string $name, string $value = '', bool $safeDefaults = true)
```
What is the reason for allowing unsafe defaults? I think we should always start with safe defaults allowing to explicitly change them to unsafe ones.

armpogart replied:
See [Cookie::fromSetCookieString](https://github.com/yiisoft/yii-web/blob/439db1b2fd5befe9a59e2315dcf2ad468f79f7a7/src/Cookie.php#L262), it creates a cookie without any defaults and later populates it from the string. If we don't have such an option, we would need to create a cookie and then chain several methods to reset the defaults while also creating several objects in the process and deleting them (all `clone` statements) and overload GC unnecessarily during that.

Also it is reasonable to allow creating minimal valid cookie without any defaults without any hassle.

armpogart replied:
@samdark I have another idea here. I'm generally against using flags anywhere (constructor, methods...). So what about if I add all available cookie attributes as constructor parameters with their defaults already set to the ones that we consider safe defaults. This will allow to more correct object construction without any flag and will allow also to construct an object with desired attributes without a need to use method chaining for building your desired object in user code, and those methods will be used when needed to modify (actually getting new object, as the class is immutable). So something like this:

```php
public function __construct(
    string $name,
    string $value = '',
    ?DateTimeInterface $expires = null,
    ?int $maxAge = null,
    ?string $domain = null,
    ?string $path = '/',
    ?bool $secure = true,
    ?string $httpOnly = true,
    ?string $sameSite = self::SAME_SITE_LAX
) {
...
}
```

samdark replied:
How about making constructor private and adding two static factories: `Cookie::create()` and `Cookie::createUnsafe()`?

armpogart replied:
@samdark Yes, that would be better, I will go that route, but I'm not convinced on naming second factory `Cookie::createUnsafe()` as technically it is just some proposed/recommended defaults and not only about safety. What about naming the second one `Cookie::createWithAttributes` or `Cookie::createWithValues` and it will accept all arguments (cookie attributes) as seen in my constructor example and users who just need custom cookie will use that.

samdark replied:
Your idea about constructor-level defaults looks good. Let's have it.

samdark commented on src/Cookie.php:
```diff
@@ -40,88 +71,120 @@ final class Cookie
+     * @var int|null maximum age of cookie in seconds
      */
-    private int $expire = 0;
+    private ?int $maxAge = null;
```
I don't think we need to store both maxAge and expire. One could be obtained from another on header generation.

As for headers it's better to always send both max age and expires (for HTTP 1.0 clients) that result in the same expiration time.

armpogart replied:
Yep, I agree. But I see one problem here. A valid cookie can have different expiry and maxAge. And according to RFC the maxAge will have precedence in that case. If we don't store maxAge and generate from expiry.
What about cookies that are not set by yii backend, cookies that we just read with `Cookie::fromSetCookieString` if we don't store maxAge and expiry separately, and cookie has both set on client side, we would just change that cookie with in an unexpected way for the user. What if some frontend or implementation somehow depends on it. I'm sure that such behavior (setting different expiry and maxAge) is wrong at all, but it is valid per RFC, and it is the main reason I'm storing maxAge.

samdark replied:
Technically it can have different expiry and maxAge but it doesn't make sense to set it like that.

> What about cookies that are not set by yii backend, cookies that we just read

We can behave as RFC is suggesting: maxAge will have precedence.

armpogart replied:
@samdark It can make sense in case if your custom client implementation somehow depends on that (not sure why).
So you propose just to set `Max-Age` anytime the expiry is set? And to also set `Expiry` to the value of  `Max-Age` when parsing the cookie from response and only `Max-Age` is there?

armpogart replied:
Also take into account that `withMaxAge` method is still useful, as it allows to just set expiry in seconds instead of creating `DateTimeInterface`, it can be renamed to `validFor` as it was before, but we will not have a way to set `Max-Age` explicitly if we remove `$maxAge` property.

samdark replied:
I don't think there's a need to make a difference between these two so:

1. Keep one value as private property.
2. Keep two methods of setting it.
3. Add both headers to response.
4. Parse incoming request according to RFC i.e. if Max-Age present, use it. Otherwise, use Expiry.

armpogart replied:
My current `maxAge` (will rename it to `withMaxAge`) method receives `DateInterval` type argument, I still would like to have a method that can also accept integer argument (seconds until expiry) as it can be handy sometimes, but I don't want to add another method for that only and also I see a use for `DateInterval` either (as it is now). The union types would (type hinting) would solve that problem, but unfortunately that RFC will land in PHP 8 by the end of year only.

Do you think we need another method (if we keep `withMaxAge` at all), or keep just this signature until union types and will update (BC break) in some future version?

armpogart replied:
or make `withMaxAge` receive integer instead of `DateInterval` for now to be consistent `Max-Age` cookie attribute that is integer?

P.S. sorry for comment spamming.

samdark replied:
Interval is enough, I think. It is convenient to use it even with seconds:

```php
$seconds = 7200;

$interval = new DateInterval('PT' . $seconds . 'S');
// or
$interval = DateInterval::createFromDateString($seconds . ' seconds');
```

samdark commented on src/Cookie.php:
```diff
@@ -40,88 +71,120 @@ final class Cookie
+     * @var string|null the path on the server in which the cookie will be available on.
      */
-    private bool $secure = true;
+    private ?string $path = null;
```
I'd make `/` default since it's what majority users expect and add `withoutPath()` method for choosing not to send it in the header.

armpogart replied:
Again, see my other answer. I just made a way to build a minimal valid cookie and am setting defaults in constructor with flag which is on by default.

samdark commented on src/Cookie.php:
```diff
@@ -40,88 +71,120 @@ final class Cookie
 
     /**
-     * @var bool whether the cookie should be accessible only through the HTTP protocol.
+     * @var bool|null whether cookie should be sent via secure connection
```
```suggestion
     * @var bool|null whether cookie should be sent via secure connection only
```

samdark commented on src/Cookie.php:
```diff
@@ -40,88 +71,120 @@ final class Cookie
+
+    public function isExpired(): bool
+    {
+        return isset($this->expire) && $this->expire->getTimestamp() < (new DateTimeImmutable())->getTimestamp();
```
```suggestion
        return $this->expire !== null && $this->expire->getTimestamp() < (new DateTimeImmutable())->getTimestamp();
```

samdark commented on src/Cookie.php:
```diff
@@ -144,37 +207,139 @@ public function httpOnly(bool $httpOnly): self
+    public static function fromSetCookieString(string $string): self
+    {
+        // array_filter with empty callback is used to filter out all falsy values
+        $rawAttributes = array_filter(preg_split('~\s*[;]\s*~', $string));
```
Can we use PREG_SPLIT_NO_EMPTY?

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+     * @param ResponseInterface $response the response object to populate from
+     * @return self collection created from response
+     */
+    public static function fromResponse(ResponseInterface $response): self
```
What is the use case for it?

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+use ArrayIterator;
+use Closure;
+use Countable;
+use DateTimeImmutable;
```
Not used.

armpogart replied:
What do you mean by not used? `CookieCoollection` implements `Countable`, it uses `ArrayIterator` in `getIterator` metho and also uses `Closure` in `exists` method.

`DateTimeImmutable` and `Psr\Http\Message\ServerRequestInterface` are not used, I just forgot to remove `use` statements. I still have some pending changes, and will remove them either.

samdark replied:
Yes, I've meant only the lines I've commented on.

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+use InvalidArgumentException;
+use IteratorAggregate;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
```
Not used.

armpogart replied:
see my other answer.

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+
+    /**
+     * Returns the collection as a PHP array.
+     * The array keys are cookie names, and the array values are the corresponding cookie objects.
```
Could be type-hinted as `@return Cookie[]`.

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+     * If there is already a cookie with the same name in the collection, it will be removed first.
+     * @param Cookie $cookie the cookie to be added
+     */
+    public function add(Cookie $cookie): void
```
What's the goal of adding a cookie to it? Is `addToResponse(ResponseInterface $response): ResponseInterface` planned?

armpogart replied:
Yep, will push the pending changes soon, they are almost ready. The class will have `addToResponse` and `setToResponse`, and it also already have `fromResponse`. So this is just a collection of cookies and easy way to handle (add, modify,...) all the cookies together when needed.

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,298 @@
+/**
+ * A CookieCollection is a collection implementation to wrap an array of Cookie class instances
+ */
+final class CookieCollection implements IteratorAggregate, ArrayAccess, Countable
```
The use case overall isn't clear.

armpogart replied:
Yep, I'm working rewriting some of the php doc blocks with more clear texts. And also adding php docs where they are absent in both classes. I'm not a native English speaker, and don't have sufficient experience writing documentation, so pardon any mistakes there.

samdark replied:
Me neither. Eventually natives may correct us. That's fine.

roxblnfk commented on src/Cookie.php:
```diff
@@ -1,14 +1,45 @@
 {
+    /**
+     * Regular Expression used to validate cookie name
+     * @see https://tools.ietf.org/html/rfc6265#section-4.1.1
```
All `@see` annotations with url to RFC should be replaced to `@link`

armpogart replied:
Yep, thanks, will push those changes soon.

armpogart commented:
Also what about naming of methods in `Cookie` class?

Do we keep them as they are `expires()`, `domain()`, `path()`... or we rename it to `withExpires()`, `withDomain()`, `withPath`... to clearly indicate that they are copying the original object?

samdark commented:
Previously cookie was used as write-only so there was no confusion about what these methods do. Now, as you're adding collections, it's a bit different and `with` may come in handy.

armpogart committed:
Additional refactoring, tests, and docs

armpogart commented:
Pushed all the changes with tests. Added all getters to `Cookie` for consistency. Removed `expire()` method from `Cookie` as it was used only in one place in `CookieCollection` which is just replaced with setting date in the past. Also I'm in doubt whether `expireWhenBrowserClosed` method needs to remain (though if we remove it, the only way to convert already existing cookie to session cookie would be to create a new one with all the same parameters except expires).

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,352 @@
+     * @return bool whether cookie exists
+     * @see has()
+     */
+    public function contains(Cookie $cookie): bool
```
Isn't it the same as `has()`?

armpogart replied:
Nope, `has()` just checks the existence by cookie name, `contains` checks the existence by value (or in this case by reference).

Also implementation of the planned interface (see answer below).

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,352 @@
+     * @param Closure $p The predicate.
+     * @return bool whether the predicate is true for at least on cookie.
+     */
+    public function exists(Closure $p): bool
```
What's the use case?

armpogart replied:
To check existence of the cookie by some attribute, e.g.:
```php
$collection->exists(fn (string $name, Cookie $cookie) => $cookie->getValue() === 'something');
```

Also an implementation of the planned interface (see answer below).

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,352 @@
+     *
+     * @return array The cookie in the collection, in the order they appear in the collection.
+     */
+    public function getCookies(): array
```
What's the use case for it?

armpogart replied:
It's just an implementation of the planned `Collection` interface (see my answer below). Though it would need to be renamed to `getValues` and `getNames` will be renamed to `getKeys`.

samdark commented on src/CookieCollection.php:
```diff
@@ -0,0 +1,352 @@
+     *
+     * @return bool whether the collection is empty.
+     */
+    public function isEmpty(): bool
```
What's the use case for it?

armpogart replied:
It's just a an implementation of the planned `Collection` interface (see my answer below).

samdark committed:
Remove expire() from collection, add it to Cookie. Some minor cleanup.

samdark commented:
I've re-added expire() to Cookie and removed it from collection.

xepozz commented:
Would be great to have a self-sufficient Collection, not CookieCollection only. 
We have yiisoft/yii2-collection. Would you like to try to port it?

samdark commented:
I think @romkatsu said he'll do it.

armpogart commented:
@samdark @xepozz The general idea with `CookieCollection` is to implement some `Collection` interface. which I was planning to extract later on to `yiisoft/arrays`. I didn't even know about existence of `yiisoft/yii2-collection` all these years working with yii2. Is there any reason to have a separate `yiisoft/collection` package for just a couple of classes and interface (it can easily fit to `yiisoft/arrays`). The `Collection` class from `yiisoft/yii2-collection` is just my planned `ArrayCollection` that implements `Collection` interface. `CookieCollection` is specialized and type-hinted case of that collection just for working with cookies. My collection interface was inspired from `CookieCollection` from yii2 and `Collection` from Doctrine collections. I will be glad to extract `Collection` interface after this PR and `ArrayCollection` implementation.

armpogart commented:
> I've re-added expire() to Cookie and removed it from collection. See #235

Do you mean to re-add `Cookie::expire` method for the mentioned PR. What about keeping also `CookieCollection::expire` in that case as it will save a few lines of code. Without it to expire just one cookie from the collection you would do:
```php
$cookie = $collection->get('someName');
$cookie->expire();
```
with it you would just do:
```php
$collection->expire('someName`);
```

As it is just a helper also in `Cookie` class, it can stay also `CookieCollection`, I just think it can be renamed `expireCookie` for clarity.

samdark commented:
> What about keeping also CookieCollection::expire in that case as it will save a few lines of code. 

It doesn't save much:

```php
// collection
$collection->expire('someName');

// cookie
$collection->get('someName')->expire();
```

armpogart commented:
> > What about keeping also CookieCollection::expire in that case as it will save a few lines of code.
> 
> It doesn't save much:
> 
> ```
> // collection
> $collection->expireCookie('someName');
> 
> // cookie
> $collection->get('someName')->expire();
> ```

This will not expire the cookie in the collection as `Cookie` is immutable, it will create a new cookie. My example is also incorrect. So to `expire` cookie in the collection we would need:
```php
$collection->set('name', $collection->get('name')->expire());
```
Or we will need to re-add `expire()`.

armpogart committed:
Fix issues, add more tests, add walk method to collection.

armpogart committed:
Fix StyleCI issue

armpogart committed:
Fix StyleCI/Scrunitezer issues

armpogart commented:
I have pushed some more changes:
- Fixed StyleCI issue.
- Fixed Scrunitizer issues.
 - Added more tests to verify the correct behavior of `CookieCollection::fromArray()` method.
- Renamed some methods and added `CookieCollection::walk()` method which helps to apply user defined function (closure) to each cookie. The usage is for batch editing cookies (later will be used when signing and/or encrypting cookies). The name of the method was chosen to be consistent with native `array_walk` and is planned in `Collection` interface.

samdark commented:
It fails tests now: https://travis-ci.com/github/yiisoft/yii-web/builds/161267353#L340

samdark commented:
About the expire... let's re-add the method to collection but keep the method in the Cookie as well.

samdark commented:
> Renamed some methods and added CookieCollection::walk() method which helps to apply user defined function (closure) to each cookie.

I'd keep collection interface to a minimum for now. @romkatsu is on general collection implementation that is reusable checking Yii 1 collections, yii2-collection and yiisoft/data to get the max out of it. Then it could be used for RBAC, this collection etc. But before it's done we're fine with current implementation.

armpogart commented:
> > Renamed some methods and added CookieCollection::walk() method which helps to apply user defined function (closure) to each cookie.
> 
> I'd keep collection interface to a minimum for now. @romkatsu is on general collection implementation that is reusable checking Yii 1 collections, yii2-collection and yiisoft/data to get the max out of it. Then it could be used for RBAC, this collection etc. But before it's done we're fine with current implementation.

Ok, I will keep an eye on that and will propose any ideas/changes later, when @romkatsu will publish the work.

I will add an `expire()` method back to `CookieCollection` to and fix the tests today.

armpogart committed:
Add CookieCollection::expire() method

samdark commented:
Good job on this one 👍
